src/uu/runcon/src/runcon.rs

```diff
@@ -0,0 +1,422 @@
+                .and_then(|_| -> Result<()> { Err(execute_command(command, &options.arguments)) })
+                {
+                    show_error!("{}.", report_full_error(&r));
+                    return libc::EXIT_FAILURE;
+                }
+            } else if let Err(r) = print_current_context() {
+                show_error!("{}.", report_full_error(&r));
+                return libc::EXIT_FAILURE;
+            }
+        }
+    }
+
+    libc::EXIT_SUCCESS
+}
+
+pub fn uu_app() -> App<'static, 'static> {
+    App::new(executable!())
+        .version(VERSION)
+        .about(ABOUT)
+        .after_help(DESCRIPTION)
+        .arg(
+            Arg::with_name(options::COMPUTE)
+                .short("c")
+                .long(options::COMPUTE)
+                .takes_value(false)
+                .help("Compute process transition context before modifying."),
+        )
+        .arg(
+            Arg::with_name(options::USER)
+                .short("u")
+                .long(options::USER)
+                .takes_value(true)
+                .value_name("USER")
+                .help("Set user USER in the target security context."),
+        )
+        .arg(
+            Arg::with_name(options::ROLE)
+                .short("r")
+                .long(options::ROLE)
+                .takes_value(true)
+                .value_name("ROLE")
+                .help("Set role ROLE in the target security context."),
+        )
+        .arg(
+            Arg::with_name(options::TYPE)
+                .short("t")
+                .long(options::TYPE)
+                .takes_value(true)
+                .value_name("TYPE")
+                .help("Set type TYPE in the target security context."),
+        )
+        .arg(
+            Arg::with_name(options::RANGE)
+                .short("l")
+                .long(options::RANGE)
+                .takes_value(true)
+                .value_name("RANGE")
+                .help("Set range RANGE in the target security context."),
+        )
+        .arg(Arg::with_name("ARG").multiple(true))
+}
+
+#[derive(Debug)]
+enum CommandLineMode {
+    Print,
+
+    PlainContext {
+        context: OsString,
+        command: OsString,
+    },
+
+    CustomContext {
+        /// Compute process transition context before modifying.
+        compute_transition_context: bool,
+
+        /// Use the current context with the specified user.
+        user: Option<OsString>,
+
+        /// Use the current context with the specified role.
+        role: Option<OsString>,
+
+        /// Use the current context with the specified type.
+        the_type: Option<OsString>,
+
+        /// Use the current context with the specified range.
+        range: Option<OsString>,
+
+        // `command` can be `None`, in which case we're dealing with this syntax:
+        // runcon [-c] [-u USER] [-r ROLE] [-t TYPE] [-l RANGE]
+        //
+        // This syntax is undocumented, but it is accepted by the GNU implementation,
```

Comment:
> I think this is documented where it says:
>> With neither context nor command, print the current security context.

It's not, because in the following syntax:
```
runcon [-c] [-u USER] [-r ROLE] [-t TYPE] [-l RANGE]
```
the context **is** specified, but the command **is not**. And yet, the GNU implementation works fine. The context is specified by the combination of the parameters `[-c] [-u USER] [-r ROLE] [-t TYPE] [-l RANGE]`.

> I think it would be cleaner if you could make parse_command_line return CommandLineMode::Print right away and make command not an option here.

Conceptually, I think of printing simply as a mode of operation of `runcon`, and not as an exceptional situation. That is why it is one of the enumerators of `CommandLineMode`. Given the way I think of it, it wouldn't be cleaner. I hope I understand your comment correctly.

Reply:
Makes sense, if you think it's clearer this way let's not change it. I interpreted it the way that `context` in the help text refers to the `context` argument in the second form, not the combination of arguments in the first form.